### PR TITLE
Context Manager Macros

### PR DESCRIPTION
**Pull request by scopatz**
This builds on #1653 and should probably be merged after that one.  This adds context manager macros to the language and is the last new bit of macro syntax I have in mind.  Basically, it lets the user totally pause parsing while within a special `with!` block. 

Using this, I am hoping to have special SSH and foreign shell blocks.  For example, something like

```
with! bashblock(login=True):
    export DATA="${DATAPATH}"
```

seems like it could be super useful.  Not to mention this idea came up a bunch in the Block PR.

### CODE REVIEW

**Comment by melund** [comment]
This is also really awesome. I can only begin to imagine the power of this...  If python were the Force, then xonsh would definitely be the dark side.

>Once you start down the xonsh path, forever will it dominate your destiny

**Comment by scopatz** [comment]
Haha

**Comment by scopatz** [comment]
OK! I think this is ready again.

**Comment by melund** [comment]
What do you think about having something like `BashBlock` (and maybe `CmdBlock`)  built into xonsh? The good thing about this is that Pygments could be made to recognize it and switch highlighter. That would be neat.

File: docs/tutorial_macros.rst [hunk 1]
```diff
@@ -427,6 +427,116 @@ really want to write some code in another language:
...
+   context manager may use them even after the object is exited. The
+   ``__exit__()`` method may clean up these attributes, if desired.
+
+By default, macro blocks are returned as a string. However, like with function
+macro arguments, the kind of ``macro_block`` is determined by a special
+annotation.  This annotation is given via the ``__xonsh_block__`` attribute
+on the context manager itself.  This allows the block to be interpreted as
+an AST, byte compiled, etc.
+
+The conveinent part about this syntax is that the macro block is only
```
Comment (melund) [inline docs/tutorial_macros.rst]: convenient

File: docs/tutorial_macros.rst [hunk 2]
```diff
@@ -427,6 +427,116 @@ really want to write some code in another language:
...
+exited once it sees a dedent back to the level of the ``with!``. All other
+code is indiscriminatly skipped! This allows you to write blocks of code in
```
Comment (melund) [inline docs/tutorial_macros.rst]: indiscriminately

File: docs/tutorial_macros.rst [hunk 3]
```diff
@@ -427,6 +427,116 @@ really want to write some code in another language:
...
+        # make sure the macro_block comes back as a string
+        __xonsh_block__ = str
+
+        def __enter__(self):
+            # parse and return the block on entry
+            root = ET.fromstring(self.macro_block)
+            return root
+
+        def __exit__(self, *exc):
+            # no reason to keep these attributes arround.
```
Comment (melund) [inline docs/tutorial_macros.rst]: arround

File: docs/tutorial_macros.rst [hunk 4]
```diff
@@ -427,6 +427,116 @@ really want to write some code in another language:
...
+And if you run this, you'll see that the ``tree`` object really is a parsed
+XML object.
+
+.. code-block:: xonshcon
+
+    >>> print(tree.tag)
+    note
+
+
+So in roughly eight lines of xonsh code, you can seemlessly interface
```
Comment (melund) [inline docs/tutorial_macros.rst]: seamlessly

File: docs/tutorial_macros.rst [hunk 5]
```diff
@@ -427,6 +427,116 @@ really want to write some code in another language:
...
+with another, vastly different language.
+
+The possibilies for this are not limited to just markup languages or other
```
Comment (melund) [inline docs/tutorial_macros.rst]: possibilities

**Comment by melund** [comment]
Only small typos. Else LGTM

**Comment by scopatz** [comment]
> What do you think about having something like BashBlock (and maybe CmdBlock) built into xonsh? The good thing about this is that Pygments could be made to recognize it and switch highlighter. That would be neat.

Yeah, I definitely want to have that, but I think that it is better implemented as a separate PR

**Comment by scopatz** [comment]
Thanks @melund! I haver taken care of the typos

**Comment by laerus** [comment]
merging this cause apparently travis is drunk again, thanks guys!

**Comment by scopatz** [comment]
Whoohoo!

**Pull request merged**